### PR TITLE
Optimize interpolation performance by pre-cropping data.

### PR DESCRIPTION
**Pull request by Asachoo**
I observed performance bottlenecks in the `Network.interpolate` function while performing computations with a large number of frequency slices/interpolations.

To address this, I've optimized the interpolation method by implementing frequency and basis slicing in advance. This approach avoids **unnecessary calculations (using the whole frequency range)** during interpolation while ensuring the same interpolation results using only **local data (wrapping the target frequency range)**.

To verify that the interpolation relies solely on local data, I conducted several tests.
```python
import time

import numpy as np

import skrf as rf

if __name__ == "__main__":
    # Set up some constants
    np.random.seed(0)
    ncalls, nfreq, nports = 1000, 5001, 20

    # Create a random network
    freq = rf.Frequency(1, 10, nfreq, "ghz")
    gen_d = lambda f, p: 2.0 * np.random.rand(f, p, p) - 1.0
    s_paras = 1.j * gen_d(nfreq, nports) + gen_d(nfreq, nports)

    ntwk = rf.Network(frequency=freq, s=s_paras)

    # Create a new frequency object
    new_freq = rf.Frequency(4, 5, 1001, "ghz")

    # Interpolate the network using the interpolate method
    ntwk_interp_direct = ntwk.interpolate(new_freq)

    # Interpolation with Pre-cropped Network
    l_idx = np.searchsorted(ntwk.f, new_freq.f[0], side="left") - 1
    r_idx = np.searchsorted(ntwk.f, new_freq.f[-1], side="right") + 1
    cropped_ntwk = ntwk[f"{ntwk.f[l_idx]}-{ntwk.f[r_idx]}hz"]
    ntwk_interp_cropped = cropped_ntwk.interpolate(new_freq)

    print(f"Is the interpolated network the same? {ntwk_interp_direct == ntwk_interp_cropped}")
```
and the result is:
```bash
Is the interpolated network the same? True
```

Your feedback and suggestions on this optimization are welcome!

### CODE REVIEW

**Comment by Asachoo** [comment]
The following benchmark demonstrates the performance improvements after modifying the `Network.interpolate()` method:
```python
import time

import numpy as np

import skrf as rf


def create_ntwk(n_f:int, n_p:int= 20):
    freq = rf.Frequency(start=10, stop=6000, npoints=n_f, unit="mhz")
    s = 1.0j * (np.random.rand(n_f, n_p, n_p) * 2.0 - 1.0)
    s += np.random.rand(n_f, n_p, n_p) * 2.0 - 1.0
    ntwk = rf.Network(frequency=freq, s=s, name="Random Ntwk")
    return ntwk


if __name__ == "__main__":
    ntimes = 100
    n_fs = [100, 500, 1000, 5000, 10000]
    
    np.random.seed(0)
    for n_f in n_fs:
        ntwk = create_ntwk(n_f, 40)
        new_freq = rf.Frequency(start=2300, stop=2400, npoints=1001, unit="mhz")
        t_list = np.zeros(ntimes)
        for i in range(ntimes):
            start = time.time()
            ntwk.interpolate(new_freq)
            end = time.time()
            t_list[i] = (end - start) * 1000.0
        print(f"n_f={n_f} on {new_freq}, mean:{t_list.mean():.2f}ms, min:{t_list.min():.2f}ms")
    print()
    for n_f in n_fs:
        ntwk = create_ntwk(n_f, 40)
        t_list = np.zeros(ntimes)
        for i in range(ntimes):
            start = time.time()
            ntwk['300MHz']
            end = time.time()
            t_list[i] = (end - start) * 1000.0
        print(f"n_f={n_f} on 300MHz, mean:{t_list.mean():.2f}ms, min:{t_list.min():.2f}ms")
```
And the output:
```bash
(skrf) (base) ➜  scikit-rf git:(TouchStoneFormat) ✗ git switch master

(skrf) (base) ➜  scikit-rf git:(master) ✗ python interpolate_performance.py     
n_f=100 on 2300.0-2400.0 MHz, 1001 pts, mean:30.69ms, min:29.39ms
n_f=500 on 2300.0-2400.0 MHz, 1001 pts, mean:33.74ms, min:32.79ms
n_f=1000 on 2300.0-2400.0 MHz, 1001 pts, mean:36.24ms, min:34.96ms
n_f=5000 on 2300.0-2400.0 MHz, 1001 pts, mean:81.91ms, min:80.36ms
n_f=10000 on 2300.0-2400.0 MHz, 1001 pts, mean:133.71ms, min:130.86ms

n_f=100 on 300MHz, mean:0.48ms, min:0.45ms
n_f=500 on 300MHz, mean:2.31ms, min:2.08ms
n_f=1000 on 300MHz, mean:5.24ms, min:4.78ms
n_f=5000 on 300MHz, mean:53.29ms, min:52.24ms
n_f=10000 on 300MHz, mean:104.37ms, min:102.35ms

(skrf) (base) ➜  scikit-rf git:(master) ✗ git switch NetworkInterpolate

(skrf) (base) ➜  scikit-rf git:(NetworkInterpolate) ✗ python interpolate_performance.py
n_f=100 on 2300.0-2400.0 MHz, 1001 pts, mean:31.64ms, min:28.81ms
n_f=500 on 2300.0-2400.0 MHz, 1001 pts, mean:30.14ms, min:29.41ms
n_f=1000 on 2300.0-2400.0 MHz, 1001 pts, mean:29.73ms, min:28.21ms
n_f=5000 on 2300.0-2400.0 MHz, 1001 pts, mean:47.28ms, min:45.68ms
n_f=10000 on 2300.0-2400.0 MHz, 1001 pts, mean:62.49ms, min:61.23ms

n_f=100 on 300MHz, mean:0.24ms, min:0.22ms
n_f=500 on 300MHz, mean:0.61ms, min:0.54ms
n_f=1000 on 300MHz, mean:1.37ms, min:1.22ms
n_f=5000 on 300MHz, mean:17.13ms, min:16.56ms
n_f=10000 on 300MHz, mean:33.25ms, min:32.46ms
```
It is evident that both **frequency slicing** and **interpolation** operations related to this PR show significant enhancements.

Additionally, according to my profiling, the operation `ntwk['300MHz']` exhibits increased execution time as the number of frequency points in the original `Network` increases. This is due to the `self.copy()` method within `Network.interpolate()` performing a copy of `ntwk.s`.
https://github.com/scikit-rf/scikit-rf/blob/5d0c93966cacb0c752d3e07e9ab484ede84a177d/skrf/network.py#L2789-L2794

**Comment by FranzForstmayr** [comment]
I'm not sure if would be an differenence if an higher order interpolation would be used.
So it's definitely a benefit, but if using cubic for example, one more point is necessary than for linear I'd assume.

**Comment by Asachoo** [comment]
> I'm not sure if would be an differenence if an higher order interpolation would be used. So it's definitely a benefit, but if using cubic for example, one more point is necessary than for linear I'd assume.

Thank you for your detailed suggestions. I updated my test cases and ran the following code in the master branch. The experiments revealed that expanding by only 1 to 2 points isn't sufficient. At least **8 points** are needed to achieve exactly the same results (I disregarded the requirement for the number of frequency points for spline interpolation).

Unfortunately, since the `rational_interp()` method uses all frequency points for rational interpolation, it's impossible to replicate the results obtained from using all frequency points by interpolating with only a subset of them.

```python3
import numpy as np

import skrf as rf

if __name__ == "__main__":
    # Set up some constants
    np.random.seed(0)
    ncalls, nfreq, nports = 1000, 5001, 20

    # Create a new frequency object
    for new_freq in (rf.Frequency(4, 4, 1, "ghz"), rf.Frequency(4, 5, 1001, "ghz")):
        for kind in (
            "linear",
            "nearest",
            "nearest-up",
            "zero",
            "slinear",
            "quadratic",
            "cubic",
            "previous",
            "next",
            "rational",
        ):
            # Create a random network
            freq = rf.Frequency(1, 10, nfreq, "ghz")
            gen_d = lambda f, p: 2.0 * np.random.rand(f, p, p) - 1.0
            s_paras = 1.0j * gen_d(nfreq, nports) + gen_d(nfreq, nports)

            ntwk = rf.Network(frequency=freq, s=s_paras)

            # Interpolate the network using the interpolate method
            ntwk_interp_direct = ntwk.interpolate(new_freq, kind=kind)

            # Interpolation with Pre-cropped Network
            boun = 8
            l_idx = np.searchsorted(ntwk.f, new_freq.f[0], side="left") - boun
            r_idx = np.searchsorted(ntwk.f, new_freq.f[-1], side="right") + boun
            l_idx = max(0, l_idx)
            r_idx = min(len(ntwk.f), r_idx)

            if kind == "rational":
                l_idx, r_idx = 0, len(ntwk.f)

            # Crop the network and interpolate again
            cropped_freq = rf.Frequency.from_f(ntwk.f[l_idx:r_idx])
            cropped_ntwk = ntwk[cropped_freq]
            ntwk_interp_cropped = cropped_ntwk.interpolate(new_freq, kind=kind)

            # record the result
            rst = ntwk_interp_direct == ntwk_interp_cropped
            print(
                f"Is the interpolated network the same with {kind} method in {new_freq}? {rst}"
            )
```

and the output,
```bash
Is the interpolated network the same with linear method in 4.0-4.0 GHz, 1 pts? True
Is the interpolated network the same with nearest method in 4.0-4.0 GHz, 1 pts? True
Is the interpolated network the same with nearest-up method in 4.0-4.0 GHz, 1 pts? True
Is the interpolated network the same with zero method in 4.0-4.0 GHz, 1 pts? True
Is the interpolated network the same with slinear method in 4.0-4.0 GHz, 1 pts? True
Is the interpolated network the same with quadratic method in 4.0-4.0 GHz, 1 pts? True
Is the interpolated network the same with cubic method in 4.0-4.0 GHz, 1 pts? False
Is the interpolated network the same with previous method in 4.0-4.0 GHz, 1 pts? True
Is the interpolated network the same with next method in 4.0-4.0 GHz, 1 pts? True
Is the interpolated network the same with rational method in 4.0-4.0 GHz, 1 pts? True
Is the interpolated network the same with linear method in 4.0-5.0 GHz, 1001 pts? True
Is the interpolated network the same with nearest method in 4.0-5.0 GHz, 1001 pts? True
Is the interpolated network the same with nearest-up method in 4.0-5.0 GHz, 1001 pts? True
Is the interpolated network the same with zero method in 4.0-5.0 GHz, 1001 pts? True
Is the interpolated network the same with slinear method in 4.0-5.0 GHz, 1001 pts? True
Is the interpolated network the same with quadratic method in 4.0-5.0 GHz, 1001 pts? True
Is the interpolated network the same with cubic method in 4.0-5.0 GHz, 1001 pts? True
Is the interpolated network the same with previous method in 4.0-5.0 GHz, 1001 pts? True
Is the interpolated network the same with next method in 4.0-5.0 GHz, 1001 pts? True
Is the interpolated network the same with rational method in 4.0-5.0 GHz, 1001 pts? True
```

**Comment by Asachoo** [comment]
1. Added a `freq_cropped` argument to the `interpolate()` method with a default value of `True`. This argument controls the frequency cropping behavior.
2. Avoid using cropped frequencies for interpolation when `kind='rational'`.
3. Updated the relevant unit test case to verify that frequency cropping behavior does not affect the interpolation results.

**Comment by FranzForstmayr** [comment]
What about to automatically crop the data except the 8 next points (to have enough for the interpolations used atm) and do not crop if `rational` is used? This would not need any user interaction.

**Comment by Asachoo** [comment]
> What about to automatically crop the data except the 8 next points (to have enough for the interpolations used atm) and do not crop if `rational` is used? This would not need any user interaction.

I agree with you. In the current PR, `freq_cropped` defaults to `True` unless explicitly set to `False` (as seen in `test_interpolate_freq_cropped()`).

```python3
...
is_rational = False
freq_cropped = kwargs.pop('freq_cropped', True)
if kwargs.get('kind', None) == 'rational':
    f_interp = mf.rational_interp
...
```

**Review by jhillairet — APPROVED** [review]

**Comment by FranzForstmayr** [comment]
@Asachoo could you change the PR as requested or do you have any thoughts against?

**Comment by Asachoo** [comment]
Thank you very much for your attention to this PR, @FranzForstmayr. I believe I have already implemented the requests for "automatically cropping the data" and "not cropping for rational interpolation" in a previous commit https://github.com/scikit-rf/scikit-rf/pull/1122/commits/88cbaf024c4f9f1b30f748e73ee29f48924c0450.

1. I have defined a variable `freq_cropped` that defaults to `True` using a `pop()` method, which will remain `True` unless the user manually sets it to `False`.
```python3
freq_cropped = kwargs.pop('freq_cropped', True)
```

2. In the if-else branch where `kind == 'rational'` is checked, I have also defined a variable `is_rational`.
```python3
is_rational = False
if kwargs.get('kind', None) == 'rational':
    f_interp = mf.rational_interp
    #Not supported by rational_interp
    del kwargs['kind']
    is_rational = True
```

3. Finally, by using the condition below, it is determined whether to crop the data. This means that when `is_rational == True` or `freq_cropped == False`, `l_idx` and `r_idx` will be reset to point to the start and end of the data, and no cropping will occur:
```python3
# rational method or prohibit frequency clipping
if is_rational or not freq_cropped:
    l_idx, r_idx = 0, len(f)
f_cropped = f[l_idx:r_idx]
```

In summary, I believe these modifications address the two requests you mentioned. If I missed any details or misunderstood your request, please let me know. Thank you once again!

**Comment by FranzForstmayr** [comment]
Sorry, missed that :)
